gigascience/gigadb-website
# Fix accessibility and styles issues in annotate files page of FUW feature

# Pull request for issue: #1698

This PR should be merged after https://github.com/gigascience/gigadb-website/pull/1730

This is a pull request for the following functionalities:

http://gigadb.gigasciencejournal.com/authorisedDataset/annotateFiles/id/200070
![image](https://github.com/gigascience/gigadb-website/assets/143437854/2b2a3954-74f4-4629-a0a3-3059514fec71)
![image](https://github.com/gigascience/gigadb-website/assets/143437854/54a97906-ea11-4fd4-8b1a-2dbe5c754919)


* Add a handful of a11y features to annotate files page
* Improve and update styles for the same page

## How to test?

* Go through the FUW workflow up to the file annotation step
  * After switching branch, you might need to run these commands to rebuild the vue widget and the css files:
    * `docker-compose run --rm less`
    * `docker-compose run --rm js sh -c 'npm install; npm run build ; npm run deploy'`
* Review the page appearance and layout
* Navigate the page with keyboard
* try to submit all of the forms with empty fields, there should be client side navigation preventing form submission
* toggle the attributes drawer
  * press tab and shift tab. Focus should be trapped within drawer
  * press "add" button with empty fields, there should be errors on display and the first empty field should receive focus
  * fill fields and press "add", and new attr should be added
  * delete attribute and press tab key: focus should remain within drawer
  * close drawer, focus should return to the button that opened it
* upload an excel for bulk upload metadata (sample id is disabled so you will need an exccel that omits that property)
* Submit metadata, and the result should be successful
* Use WAVE extension to review a11y issues

## How have functionalities been implemented?

* Accessibility issues
  * submit button is displayed in a disabled state rather than hidden. The text content was updated to reflect this change.
  * fixed any color contrast issues
  * add aria-label to delete buttons
  * Page uses modern layout, i.e. with "skip to main" button
  * Added labels to inputs that missed them, the aria-labelledby attribute was used to reference row + column name
* Forms
  * Previously the entire content was wrapped in one single form, which had nested forms as children. I rearranged the html to avoid having nested forms, so now the form that submits the files and metadata, the form that bulk uploads with an excel file and the form to add attributes (inside the drawer) are all separate forms
    * Since the html was rearranged to avoid form nesting, the submit button "Complete and return to Your Uploaded Datasets page" is now directly underneath the table. It could be repositioned at the bottom again using CSS, but didn't seem necessary, and from an accessibility perspective it would probably not be advisable to position the submit button far away from the inputs it affects
  * Added client side validation: now the data type and description fields are mandatory and must be filled before form submission
  * Empty table headers got label (Actions)
  * Attributes drawer is in separate component and wrapped in a focus trap that adds a11y features typical of modals:
    * trap focus within when drawer is open
    * return focus to element that triggered drawer, on close
  * Attributes drawer got a more customized heading using a Vue slot, this allows to specify the heading level
  * Added form validation to drawer inputs
* Styles
  * improved alignment of page elements
  * increased width of tips panel to make it look a bit less cramped
  * reduced inner padding of box to bulk upload metadata
  * used updated brand styles for all buttons
  * tips boxes styles were updated
* Other
  * Page now uses TitleBreadcrumb component
  * attributes drawer scrolling behavior was changed. Instead of scrolling only the table content and give the table a fixed height, the table height is dynamic and the entire drawer body is scrollable. this minimizes empty space within the content. Also, keeping the scroll to the table caused a strange layout shift on the delete buttons
  * Refactored the top page notifications in protected/views/authorisedDataset/filesAnnotate.php

NOTES:

* if user submits an invalid excel file, e..g mismatching properties, there is no user feedback. I think it would be useful to show an alert on the top of the page if that is the case
* I noticed the role="alert" usage is not as theoretically intended (notifiy user on dynamic children, not for static content rendered on page load), so this would involve some refactoring as the `hasFlash` method does not seem to render the content dynamically (created an issue describing this #1729)

## Any issues with implementation?

--

## Any changes to automated tests?

--
